File: src/cpp/message.cpp
#include "message.h"

#include <string>
#include <sstream>

#include <boost/algorithm/string.hpp>

#include <iostream>

using namespace bjcomm;
using namespace boost::algorithm;

Message::Message():
    _type("default"), _data("") {}

Message::Message(std::string type):
    _type(type), _data("")
{
    trim(_type);
    if(_type.empty()) _type = "default";
}

Message::Message(std::string type, std::string data): 
    _type(type), _data(data)
{
    trim(_type);
    if(_type.empty()) _type = "default";
    trim(_data);
}
    
Message::~Message(){}

Message::Message(const Message &msg): _type(msg._type), _data(msg._data) {}

Message &Message::operator=(const Message &msg) {
    if (this == &msg) return *this; 
    _type = msg._type;
    _data = msg._data;
    return *this;
}

void Message::setType(std::string type){
    trim(type);
    if(type.empty()) type = "default";
    _type = type;
}

std::string Message::getType() const{
    return _type;
}    
    
void Message::setData(std::string data){
    trim(data);
    _data = data;
}

std::string Message::getData() const{
    return _data;
}

std::istringstream &Message::getStream(){
    //cannot pass direct copy because of gcc bug
    _stream_saver.clear();
    _stream_saver.str(_data);
    return _stream_saver;
}

void Message::clear(){
    _data.clear();
}

File: src/java/Message.java
/*
 * Copyright (c) 2015 Blue Jay Eindhoven
 */
package nl.bluejayeindhoven.bjcomm;

import java.util.Scanner;

/**
 * Data class that contains a message with a type that can be send and received
 *
 * <p>This class makes sure that the message follow the defined format
 * A message should start with an non-empty string (printable ASCII) that defines the type (default if not set).
 * Then comes one space followed by any non-space character that starts the data. The end of the data is a printable character</p>
 *
 * @author Koen Wolters
 */
public class Message{
    /**
     * Type of the message
     */
    private String type;
    /**
     * Data of the message
     */
    private String data;
    
    /**
     * Initialize a default message
     */
    public Message(){
        this.type = "default";
        this.data = "";
    }
    /**
     * Initialize message with a type
     *
     * @param type type of the message
     */
    public Message(String type){
        this.type = type.trim();
        if(type.isEmpty()) type = "default";
        this.data = "";
    }
    /**
     * Initialize message with a type and data
     *
     * @param type type of the message
     * @param data data of the message
     */
    public Message(String type, String data){
        this.type = type.trim();
        if(type.isEmpty()) type = "default";
        this.data = data.trim();
    }
    
    /**
     * Set the type of the message 
     *
     * @param type the type of the message (becomes default if type is empty string)
     */
    public void setType(String type){
        type = type.trim();
        if(type.isEmpty()) type = "default";
        this.type = type;
    }
    /**
     * Get the type of the message
     *
     * @return the type of the message
     */
    public String getType(){
        return type;
    }
    
    /**
     * Set the data of the message as string
     *
     * @param data the data of the message
     */
    public void setData(String data){
        data = data.trim();
        this.data = data;
    }
    
    /**
     * Get the data of the message as string
     *
     * @return the data of the message
     */
    public String getData(){
        return data;
    }
    
    /**
     * Returns a Scanner object that can be used to read the data. Does not modify internal data
     *
     * @return object to read the message data 
     */
    public Scanner getScanner(){
        return new Scanner(data);
    }
    
    /**
     * Clears the message data
     */
    public void clear(){
        data = "";
    }
}